src/consumption.cpp

```diff
@@ -101,69 +101,150 @@ int player::stomach_capacity() const
-        }
-        kcal /= comest.recipe_charges;
-    } else {
-        kcal = comest.get_comestible()->get_calories();
+    // As float to avoid rounding too many times
+    float kcal = comest.get_comestible()->default_nutrition.kcal;
 
-        // Many raw foods give less calories, as your body has expends more energy digesting them.
-        // We don't want RAW to stack for components and results, so we're doing it in this else block.
-        if( comest.has_flag( "RAW" ) && !comest.has_flag( "COOKED" ) ) {
-            kcal *= 0.75f;
-        }
+    // Many raw foods give less calories, as your body has expends more energy digesting them.
+    if( comest.has_flag( "RAW" ) && !comest.has_flag( "COOKED" ) ) {
+        kcal *= 0.75f;
     }
 
-    if( has_trait( trait_GIZZARD ) ) {
+    if( p.has_trait( trait_GIZZARD ) ) {
         kcal *= 0.6f;
     }
 
-    if( has_trait( trait_CARNIVORE ) && comest.has_flag( flag_CARNIVORE_OK ) &&
+    if( p.has_trait( trait_CARNIVORE ) && comest.has_flag( flag_CARNIVORE_OK ) &&
         comest.has_any_flag( carnivore_blacklist ) ) {
         // TODO: Comment pizza scrapping
         kcal *= 0.5f;
     }
 
     const float relative_rot = comest.get_relative_rot();
     // Saprophages get full nutrition from rotting food
-    if( relative_rot > 1.0f && !has_trait( trait_SAPROPHAGE ) ) {
+    if( relative_rot > 1.0f && !p.has_trait( trait_SAPROPHAGE ) ) {
         // everyone else only gets a portion of the nutrition
         // Scaling linearly from 100% at just-rotten to 0 at halfway-rotten-away
         const float rottedness = clamp( 2 * relative_rot - 2.0f, 0.1f, 1.0f );
         kcal *= ( 1.0f - rottedness );
     }
 
     // Bionic digestion gives extra nutrition
-    if( has_bionic( bio_digestion ) ) {
+    if( p.has_bionic( bio_digestion ) ) {
         kcal *= 1.5f;
     }
 
     return static_cast<int>( kcal );
 }
 
+// list of traits the player has that modifies vitamin absorption
+static std::list<trait_id> mut_vitamin_absorb_modify( const player &p )
+{
+    std::list<trait_id> traits;
+    for( auto &m : p.get_mutations() ) {
+        const auto &mut = m.obj();
+        if( !mut.vitamin_absorb_multi.empty() ) {
+            traits.push_back( m );
+        }
+    }
+    return traits;
+}
+
+// is the material associated with this item?
+static bool material_exists( const material_id &material, const item &item )
+{
+    for( const material_id &mat : item.type->materials ) {
+        if( mat == material ) {
+            return true;
+        }
+    }
+    return false;
+}
+
+static std::map<vitamin_id, int> compute_default_effective_vitamins( const item &it,
+        const player &p )
+{
+    std::map<vitamin_id, int> res;
+
+    assert( it.get_comestible() );
+
+    // if we're here, whatever is returned is going to be based on the item's defined stats
+    res = it.get_comestible()->default_nutrition.vitamins;
+    std::list<trait_id> traits = mut_vitamin_absorb_modify( p );
```

Comment:
Why is the content of `mut_vitamin_absorb_modify` in a separate function? It is only used here, so you could move the iteration over `p.get_mutations()` here. This would avoid creating a temporary vector `traits` and gathering the mutation object twice (here and in `mut_vitamin_absorb_modify`).